modules/fundamental/src/purl/service/mod.rs

```diff
@@ -230,41 +232,90 @@ impl PurlService {
         }
     }
 
-    pub async fn purl_by_purl<C: ConnectionTrait>(
+    #[instrument(skip(self, connection), err(level=tracing::Level::INFO))]
+    pub async fn fetch_purl_details<C: ConnectionTrait>(
         &self,
-        purl: &Purl,
+        identifiers: &[String],
+        deprecated: Deprecation,
+        connection: &C,
+    ) -> Result<HashMap<String, PurlDetails>, String> {
+        let mapped: Result<Vec<_>, String> = identifiers
+            .into_iter()
+            .map(|key| {
+                if key.starts_with("pkg:") {
+                    Purl::from_str(key)
+                        .map(|purl| Either::Left(purl))
+                        .map_err(|e| format!("Invalid purl '{}': {}", key, e))
+                } else {
+                    Uuid::from_str(key)
+                        .map(|uuid| Either::Right(uuid))
+                        .map_err(|e| format!("Invalid UUID '{}': {}", key, e))
+                }
+            })
+            .collect();
+        let (purls, uuids) = mapped?.into_iter().partition_map(|either| either);
+
+        let purl_details = self
+            .purls_by_purl(purls, deprecated, connection)
+            .await
+            .map_err(|e| format!("Failed to fetch purl details by purl: {}", e))?;
+        let purls = purl_details
+            .into_iter()
+            .map(|detail| (detail.head.purl.to_string(), detail));
+
+        let uuid_details = self
+            .purls_by_uuid(uuids, deprecated, connection)
+            .await
+            .map_err(|e| format!("Failed to fetch purl details by uuid: {}", e))?;
+        let uuids = uuid_details
+            .into_iter()
+            .map(|detail| (detail.head.uuid.to_string(), detail));
+
+        let result: HashMap<String, PurlDetails> = purls.chain(uuids).collect();
+        Ok(result)
+    }
+
+    async fn purls_by_purl<C: ConnectionTrait>(
+        &self,
+        purls: Vec<Purl>,
```

Comment:
It doesn't look like you actually need (to own) a `Vec` here.

You can either use `&[Purl]`, or consume the `purls` later, using:

```rust
        let canonical: Vec<CanonicalPurl> = purls
            .into_iter()
            .map(|purl| CanonicalPurl::from(purl))
            .collect();
```